### PR TITLE
[#9804] Add in masquerade mode

### LINKED ISSUE #9804: Add in masquerade mode
**Description of feature/enhancement**
Currently, in V7, the masquerade mode is not added.
Allow admin to impersonate as instructors or students to view and edit data.

**Possible solutions**
1. In each page, when need to send http request, first check if there is `user` present in `queryParams` in the url, if there is, then add `user: this.user` to the request's paramsMap, then send this masquerade request, otherwise, send the normal request.
2. add an optional parameter `user?` for HttpRequestService's get, put, update and delete method, and build the HttpParams there as in solution 1.
3. add a MasqueradeModeService to handle masquerade requests, but it still requires existence checks of `user` in each page.

May I have some suggestions for better solutions? @wkurniawan07 @xpdavid

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #9804

Some code is adopted from #9813, thanks for the contribution @crizyli .

**Outline of Solution**

- Use `MasqueradeModeService` to hold user to masquerade (stateful service like a redux store).
- Update `MasqueradeModeService` based on the auth info returned from the backend.
- Define new method `openNewWindow` in `NavigationService` so that masquerading userId get passed.
- Listen to router's `NavigationEnd` event and update `user` URL param so that URL is always consistent with the interval state of `MasqueradeModeService`.
- Remove most of `user` field in many pages

### CODE REVIEW

File: src/web/services/navigation.service.ts [hunk 1]
```diff
@@ -40,4 +42,16 @@ export class NavigationService {
       this.statusMessageService.showSuccessMessage(message);
     });
   }
+
+  /**
+   * Opens a new browser window.
+   */
+  openNewWindow(urlStr: string): void {
+    const masqueradeUser: string = this.masqueradeModeService.getMasqueradeUser();
```
Comment (crizyli) [inline src/web/services/navigation.service.ts]: Maybe can delete this line, and put `isInMasqueradingMode()` in the if statement below.

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.ts [hunk 2]
```diff
@@ -151,7 +145,7 @@ export class InstructorHomePageComponent extends InstructorSessionModalPageCompo
     this.courseService.binCourse(courseId).subscribe((course: Course) => {
       this.loadCourses();
       this.statusMessageService.showSuccessMessage(
-        `The course ${course.courseId} has been deleted. You can restore it from the Recycle Bin manually.`);
+          `The course ${course.courseId} has been deleted. You can restore it from the Recycle Bin manually.`);
```
Comment (crizyli) [inline src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.ts]: Extra indentation here?

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts [hunk 3]
```diff
@@ -192,11 +190,11 @@ export class InstructorSessionEditPageComponent extends InstructorSessionBasePag
         intent: Intent.FULL_DETAIL,
       };
       this.httpRequestService.get('/session', paramMap)
-        .subscribe((feedbackSession: FeedbackSession) => {
-          this.sessionEditFormModel = this.getSessionEditFormModel(feedbackSession);
-        }, (resp: ErrorMessageOutput) => {
-          this.statusMessageService.showErrorMessage(resp.error.message);
-        });
+          .subscribe((feedbackSession: FeedbackSession) => {
```
Comment (crizyli) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts]: Same indentation issue.

File: src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts [hunk 4]
```diff
@@ -471,9 +469,9 @@ export class InstructorSessionEditPageComponent extends InstructorSessionBasePag
    */
   private loadResponseStatusForQuestion(model: QuestionEditFormModel): void {
     this.feedbackSessionsService.hasResponsesForQuestion(model.feedbackQuestionId)
-      .subscribe((resp: HasResponses) => {
-        model.isQuestionHasResponses = resp.hasResponses;
-      }, (resp: ErrorMessageOutput) => { this.statusMessageService.showErrorMessage(resp.error.message); });
+        .subscribe((resp: HasResponses) => {
```
Comment (crizyli) [inline src/web/app/pages-instructor/instructor-session-edit-page/instructor-session-edit-page.component.ts]: Same here.

File: src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.ts (hunk 2 above)
Comment (xpdavid) [reply src/web/app/pages-instructor/instructor-home-page/instructor-home-page.component.ts]: Continuation Indentation should be 4. The previous indentation is kind of wrong.

File: src/web/services/navigation.service.ts (hunk 1 above)
Comment (xpdavid) [reply src/web/services/navigation.service.ts]: Yup

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 5]
```diff
@@ -79,6 +69,8 @@ export class StudentProfilePageComponent implements OnInit {
       if (auth.user) {
         this.id = auth.user.id;
 
+        this.profilePicLink = `${this.backendUrl}/webapi/student/profilePic?user=${this.id}`;
```
Comment (wkurniawan07) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: I assume masquerade mode will correctly be triggered here?

File: src/web/app/pages-student/student-profile-page/student-profile-page.component.ts [hunk 6]
```diff
@@ -128,17 +120,14 @@ export class StudentProfilePageComponent implements OnInit {
...
 
-      this.httpRequestService.post('/student/profilePic', paramsMap, formData)
+      this.httpRequestService.post('/student/profilePic', {}, formData)
           .subscribe(() => {
             this.statusMessageService.showSuccessMessage('Your profile picture has been saved successfully');
 
             // force reload
             const timestamp: number = (new Date()).getTime();
-            this.profilePicLink = `${this.backendUrl}/webapi/student/profilePic?${timestamp}`;
+            this.profilePicLink = `${this.backendUrl}/webapi/student/profilePic?${timestamp}&user=${this.id}`;
```
Comment (wkurniawan07) [inline src/web/app/pages-student/student-profile-page/student-profile-page.component.ts]: Similarly, here?

**Pull request merged**